Ical.Net/DataTypes/PeriodList.cs

```diff
@@ -73,57 +100,116 @@ public static Dictionary<string, List<Period>> GetGroupedPeriods(IList<PeriodLis
 
-                if (!grouped.ContainsKey(actualBucket))
+                if (!grouped.TryGetValue(bucketTzId, out var periods))
                 {
-                    grouped.Add(actualBucket, new HashSet<Period>());
+                    periods = new HashSet<Period>();
+                    grouped.Add(bucketTzId, periods);
                 }
-                grouped[actualBucket].Add(period);
+
+                periods.Add(period);
             }
         }
-        return grouped.ToDictionary(k => k.Key, v => v.Value.OrderBy(d => d.StartTime).ToList());
+
+        return grouped.ToDictionary(k => k.Key, v => (IList<Period>) v.Value.OrderBy(d => d.StartTime).ToList());
     }
 
-    protected bool Equals(PeriodList other) => string.Equals(TzId, other.TzId, StringComparison.OrdinalIgnoreCase)
-                                               && CollectionHelpers.Equals(Periods, other.Periods);
+    protected bool Equals(PeriodList other) => CollectionHelpers.Equals(Periods, other.Periods);
 
-    public override bool Equals(object obj)
+    /// <inheritdoc/>
+    public override bool Equals(object? obj)
     {
         if (ReferenceEquals(null, obj)) return false;
         if (ReferenceEquals(this, obj)) return true;
         return obj.GetType() == GetType() && Equals((PeriodList) obj);
     }
 
-    public override int GetHashCode()
-    {
-        unchecked
-        {
-            var hashCode = TzId?.GetHashCode() ?? 0;
-            hashCode = (hashCode * 397) ^ CollectionHelpers.GetHashCode(Periods);
-            return hashCode;
-        }
-    }
+    /// <inheritdoc/>
+    public override int GetHashCode() => CollectionHelpers.GetHashCode(Periods);
 
+    /// <inheritdoc/>
     public Period this[int index]
     {
         get => Periods[index];
         set => Periods[index] = value;
     }
 
+    /// <inheritdoc/>
     public bool Remove(Period item) => Periods.Remove(item);
+
+    /// <inheritdoc/>
     public bool IsReadOnly => Periods.IsReadOnly;
+
+    /// <inheritdoc/>
     public int IndexOf(Period item) => Periods.IndexOf(item);
+
+    /// <inheritdoc/>
     public void Insert(int index, Period item) => Periods.Insert(index, item);
+
+    /// <inheritdoc/>
     public void RemoveAt(int index) => Periods.RemoveAt(index);
-    public void Add(Period item) => Periods.Add(item);
+
+    /// <summary>
+    /// Adds a <see cref="Period"/> to the list.<br/>
+    /// The timezone period kind of the first value added determines the timezone for the whole list.
+    /// <para/>
+    /// Use <see cref="Collections.ExceptionDateCollection"/> and <see cref="Collections.RecurrencePeriodCollection"/>
+    /// to simplify creating 'EXDATE' and 'RDATE' properties.
+    /// </summary>
+    /// <param name="item">The <see cref="Period"/> for an 'RDATE'.</param>
+    /// <exception cref="ArgumentException"></exception>
+    public void Add(Period item)
+    {
+        EnsureConsistentTimezoneAndPeriodKind(item);
```

Comment:
I feel this validation is somewhat hopeless at this time, because there are several other places to verify (Insert, Indexer, CopyTo), but even if we tackle all of them, Period and IDateTime are still mutable, so they could still be changed (another reason why it would be great to have them immuntable).

A question that comes to my mind is whether it would be a viable alternative to allow mixing different kinds altogether, having just a single list in `CalendarEvent` and split into multiple just during serialization? We wouldn't preserve the original structure after a deserialization/serialization cycle, but that should be ok as the semantics would stay unchanged. Would also relieve us from some 'GetGrouped...' kind of code.